### ISSUE TITLE
gradient explosion while training

### ISSUE THREAD

**Issue by SHUGOSHAx** [hellopipu/PromptMR-plus#7]
Hi @hellopipu 

First, thank you for sharing your fantastic PromptMR+ work.

i found that when i train on promptmr+，after about 200 steps of training each time, there would always be a gradient explosion, causing the loss value to become 1 and no longer decrease. Have you ever encountered this situation? How can i avoid it?

Thanks.

**Comment by hellopipu** [comment]
Feel free to reopen if you have further questions.

**Comment by 5o1** [comment]
> Hi @SHUGOSHAx,
> 
> This is a known issue—I’ve also heard about it from several others using this repo.
> 
> When I released the code, I didn’t have an effective solution for it. My workaround was to add [this line](https://github.com/hellopipu/PromptMR-plus/blob/72df5d44f63672ee6c5449d5bb1703e5573a7d94/pl_modules/promptmr_module.py#L183) to detect NaN loss and simply (change the seed and) resume training. You can also modify the code to detect when the loss value is 1.
> 
> Bingyu
> 
> 
> 

The gradient explosion is because this model does NOT use ANY normalization layer. (Only one manual normalization before entre to Unet is used.)

In the varnet provided by fastmri-repo, InstanceNorm is used to avoid this issue. And the price is that convergence will become very very slow. For example, under good initialization conditions, the approach of this repo can quickly converge the training loss to the order of 1e-2. But in the normalization network, it takes several epochs to do it.
